java-checks/src/main/java/org/sonar/java/checks/serialization/SerializableFieldInSerializableClassCheck.java

```diff
@@ -54,14 +60,46 @@ public void visitNode(Tree tree) {
     if (isSerializable(classTree)
       && !SerializableContract.hasSpecialHandlingSerializationMethods(classTree)
       && !classTree.symbol().type().isSubtypeOf("javax.servlet.http.HttpServlet")) {
-      for (Tree member : classTree.members()) {
-        if (member.is(Tree.Kind.VARIABLE)) {
-          checkVariableMember((VariableTree) member);
-        }
-      }
+      
+      Set<String> constructorInjectedParams = getConstructorInjectedFields(classTree);
+      Set<String> setterInjectedParams = getSetterInjectedFields(classTree);
+      
+      classTree.members().stream()
+        .filter(member -> member.is(Tree.Kind.VARIABLE))
+        .map(VariableTree.class::cast)
+        .filter(variableTree -> 
+          !(constructorInjectedParams.contains(variableTree.simpleName().name()) ||
+            setterInjectedParams.contains(variableTree.simpleName().name())))
+        .forEach(this::checkVariableMember);
     }
   }
 
+  private static Set<String> getSetterInjectedFields(ClassTree classTree) {
+    return classTree.members().stream()
+      .filter(member -> member.is(Tree.Kind.METHOD))
+      .map(MethodTree.class::cast)
+      .filter(SerializableFieldInSerializableClassCheck::isAnnotatedWithInject)
+      .map(methodTree -> methodTree.simpleName().name())
+      .filter(name -> name.startsWith("set"))
+      .map(name -> name.substring(3).toLowerCase(Locale.ROOT))
+      .collect(Collectors.toSet());
+  }
+
+  private static Set<String> getConstructorInjectedFields(ClassTree classTree) {
+    return classTree.members().stream()
+      .filter(member -> member.is(Tree.Kind.CONSTRUCTOR))
+      .map(MethodTree.class::cast)
+      .filter(SerializableFieldInSerializableClassCheck::isAnnotatedWithInject)
+      .flatMap(methodTree -> methodTree.parameters().stream())
+      .map(variableTree -> variableTree.simpleName().name())
```

Comment:
I'm wondering if we can really rely on the naming of parameters here. As an example, we will not support the example described in the ticket.

What is your opinion on it?

Reply:
Yes, that's true. I was just thinking what would be a better option.
So I changed the logic here to check assignments inside the constructor, and exclude fields that are assigned values inside constructor annotated with @Inject.
 Wdyt about this solution?

Reply:
Nice one, this solution looks better to me! What about using it for the setter as well? Doing both in this method?

Reply:
I thought about it. "Setters" are actually a very common practice and initially I thought it will be much faster to just rely on naming convention, but you're right <code>@Inject</code> can be used on any method and it will be nice to reduce noise for them as well